Gap-7: Functions Diagnose (B → A) — Trigger + Durable troubleshooting (#1641)
microsoft/GitHub-Copilot-for-Azure: plugin/skills/azure-diagnostics/references/functions/durable.md

# Durable Functions Troubleshooting

## Common Issues Matrix

| Symptom | Likely Cause | Quick Fix |
|---------|--------------|-----------|

> | Orchestration stuck in "Running" | Activity function failed or hung | Check task hub history; terminate and purge if needed |

Reviewer: **[LOW]** `IDurableOrchestrationContext` is in-process model only. The isolated worker model (now the recommended path) uses `TaskOrchestrationContext`.

```suggestion
| Non-deterministic orchestration error | Code uses `DateTime.Now`, `Guid.NewGuid()`, or random | Replace with deterministic context APIs (`IDurableOrchestrationContext` in-process / `TaskOrchestrationContext` isolated) |
```

| Task hub conflicts | Multiple apps sharing same hub | Set unique `hubName` per app in host.json |
| Fan-out never completes | One activity silently failed | Query instance status for sub-orchestrations; check `exceptions` table |
| Replay causes side effects | I/O in orchestrator function | Move all I/O into activity functions |

---

## Stuck Orchestrations

**Diagnose:**
```bash
# Query orchestration instances by status
func durable get-instances --connection-string-setting AzureWebJobsStorage \
  --task-hub-name TASKHUB --runtime-status Running --top 10

# Check specific instance history
func durable get-runtime-status --id INSTANCE_ID \
  --connection-string-setting AzureWebJobsStorage \
  --task-hub-name TASKHUB --show-history
```

**KQL — Stuck orchestrations:**
```kql
traces
| where timestamp > ago(24h)
| where message contains "orchestration" and
  (message contains "stuck" or message contains "timeout" or message contains "failed")
| project timestamp, operation_Name, message
| order by timestamp desc
```

**Fix — Terminate and purge:**
```bash
# Terminate a stuck instance

> func durable terminate --id INSTANCE_ID --reason "Manual termination — stuck" \

Reviewer: The `--reason` value contains a Unicode em dash (U+2014). Some terminals and shells (CMD, older PowerShell, certain SSH clients) don't handle this well in copy-paste, potentially causing encoding errors or producing a garbled reason string in the orchestration metadata. Consider using a plain hyphen:

```suggestion
func durable terminate --id INSTANCE_ID --reason "Manual termination - stuck" \
```

```

> ⚠️ **Warning:** Termination is immediate and non-recoverable. Activity functions already running will complete, but their results are discarded.

---

## Non-Deterministic Code Detection

Orchestrator functions replay from history. Any non-deterministic call breaks replay.

| Anti-Pattern | Why It Breaks | Correct Alternative |
|-------------|---------------|---------------------|
| `DateTime.Now` / `DateTime.UtcNow` | Different value on replay | `context.CurrentUtcDateTime` |
| `Guid.NewGuid()` | Different GUID on replay | `context.NewGuid()` |
| `Thread.Sleep` / `Task.Delay` | Not replay-safe; uses non-durable timers so waits aren't persisted or deterministic | `context.CreateTimer()` |
| Direct HTTP calls | Different response on replay | Use activity function or `context.CallHttpAsync()` |
| Environment variables | May change between replays | Pass config as orchestrator input |
| Random number generation | Non-deterministic | Generate in activity, pass to orchestrator |

**KQL — Detect non-deterministic errors:**
```kql
exceptions
| where timestamp > ago(24h)
| where type contains "NonDeterministic" or
  outerMessage contains "non-deterministic" or

>   outerMessage contains "orchestrator function completed with"

Reviewer 2: This outerMessage condifition feels very generic. Is it guaranteed to be related to non-deterministic errors?

Reviewer: Fair question. The pattern targets the Durable Tasks framework error messages like "The orchestrator function X completed with a different result than expected" and "completed with a different task count." Those specific messages are non-determinism symptoms. That said, it's a substring match and could theoretically catch unrelated traces if custom logging happens to include that phrase. Adding a brief inline note (e.g., '-- cross-reference with type contains NonDeterministic for precision') would reduce false positives for users with verbose custom logging.

**Diagnose:**
```bash
# Check if a task hub name override is set via app settings
az functionapp config appsettings list -n APP -g RG \
  --query "[?name=='AzureFunctionsJobHost__extensions__durableTask__hubName']" -o table

# If no override is set, the hub name comes from host.json:
#   - Inspect host.json in your source repo, or
#   - Use Kudu (Advanced Tools) / zip-deployed content to view host.json in the deployed app
```

**Fix — Set unique hub names in host.json:**
```json
{
  "version": "2.0",
  "extensions": {
    "durableTask": {
      "hubName": "MyAppTaskHubProd"
    }
  }
}
```

| Symptom | Cause | Fix |
|---------|-------|-----|
| Orchestrations appear in wrong app | Shared `hubName` and storage account | Set unique `hubName` per app |
| Phantom instances after redeploy | Old hub data persists | Purge history or use new hub name |
| `PartitionNotFoundException` | Hub tables corrupted or deleted | Delete and recreate hub tables in storage |

**Storage tables for a task hub (Azure Storage provider):**
```bash
# List task hub tables
az storage table list --account-name STORAGE \
  --query "[?starts_with(name, 'TASKHUB')]" --output table
```

> > 💡 **Tip:** For the Azure Storage provider, choose a task hub name that satisfies the full provider constraints—not just “alphanumeric only.” In practice, use letters and numbers only (no hyphens, underscores, or other special characters), keep the name within the documented length limit, and ensure it starts with a letter when required by your Durable Functions runtime/provider combination. The task hub name is reused as a prefix for multiple tables and queues, including `<HubName>History`, `<HubName>Instances`, and related control/work-item queues. See the official [Durable Functions storage provider documentation](https://learn.microsoft.com/azure/azure-functions/durable/durable-functions-storage-providers#azure-storage) for the current authoritative naming rules and schema details.

Reviewer: Optional: the opening clause 'not just "alphanumeric only"' implies the character rules go beyond alphanumeric, but the practical advice that follows IS 'letters and numbers only.' The extra constraints (length limit, starts-with-letter) are what make it 'not just alphanumeric,' but they get buried mid-sentence.

Consider leading with all three rules directly:

```suggestion
> 💡 **Tip:** For the Azure Storage provider, task hub names must satisfy three constraints: letters and numbers only (no hyphens, underscores, or special characters), the name must start with a letter, and it must stay within the documented length limit for your runtime/provider. The hub name becomes a prefix for multiple tables and queues (`<HubName>History`, `<HubName>Instances`, and related control/work-item queues). See the official [Durable Functions storage provider documentation](https://learn.microsoft.com/azure/azure-functions/durable/durable-functions-storage-providers#azure-storage) for the current authoritative naming rules.
```

Author: Reworded — the tip now leads with the three naming constraints, then explains the table/queue prefixes, and links the Durable Functions storage provider docs. In 7e4cdf4.

```bash
# List task hub tables
az storage table list --account-name STORAGE \
  --query "[?starts_with(name, 'TASKHUB')]" --output table
```

> 💡 **Tip:** For the Azure Storage provider, task hub names must satisfy three constraints: letters and numbers only (no hyphens, underscores, or special characters), the name must start with a letter, and it must stay within the documented length limit for your runtime/provider. The hub name becomes a prefix for multiple tables and queues (`<HubName>History`, `<HubName>Instances`, and related control/work-item queues). See the official [Durable Functions storage provider documentation](https://learn.microsoft.com/azure/azure-functions/durable/durable-functions-storage-providers#azure-storage) for the current authoritative naming rules.

---

## Timer / Retry Pattern Issues

| Symptom | Cause | Fix |
|---------|-------|-----|
| Timer fires immediately on replay | Using `Task.Delay` instead of durable timer | Replace with `context.CreateTimer(fireAt, cancellationToken)` |
| Retry storms | `maxNumberOfAttempts` too high with short intervals | Use exponential backoff: `CallActivityWithRetryAsync` with `RetryOptions` |

> | Sub-orchestration timeout | No timeout set on `CallSubOrchestratorAsync` | Wrap in `Task.WhenAny` with `context.CreateTimer` as deadline |

Reviewer 3: `Task.WhenAny` needs a cancellation path that this guidance omits. When `CallSubOrchestratorAsync` completes first, the durable timer remains outstanding and keeps the parent orchestration in `Running` until the deadline. When the timer wins, the sub-orchestration also keeps running. Show canceling the timer on the success path, and state that a timed-out sub-orchestration must be terminated explicitly when that is required.
